lsp/nls/src/requests/completion.rs

```diff
@@ -202,62 +205,86 @@ fn remove_duplicates(items: &Vec<CompletionItem>) -> Vec<CompletionItem> {
-    id: usize,
-    path: &mut Vec<Ident>,
-) -> Vec<(Vec<Ident>, Types)> {
+fn collect_record_info(linearization: &Completed, id: usize, path: &mut Vec<Ident>) -> Vec<Ident> {
     linearization
-        .get_in_scope(item)
-        .iter()
-        .filter_map(|item| {
+        .get_item(id)
+        .map(|item| {
             let (ty, _) = linearization.resolve_item_type_meta(item);
             match (&item.kind, ty) {
                 // Get record fields from static type info
-                (TermKind::Declaration(..), Types(TypeF::Record(rrows))) if id == item.id => {
-                    Some((find_fields_from_type(&rrows, path), item.ty.clone()))
+                (TermKind::Declaration(..), Types(TypeF::Record(rrows))) => {
+                    find_fields_from_type(&rrows, path)
                 }
-                (TermKind::Declaration(_, _, ValueState::Known(body_id)), _) if id == item.id => {
-                    match find_fields_from_contract(linearization, *body_id, path) {
-                        // Get record fields from contract metadata
-                        Some(fields) => Some((fields, item.ty.clone())),
-                        // Get record fields from lexical scoping
-                        None => find_fields_from_term_kind(linearization, id, path)
-                            .map(|idents| (idents, item.ty.clone())),
-                    }
+                (TermKind::Declaration(_, _, ValueState::Known(body_id)), _) => {
+                    find_fields_from_contract(linearization, *body_id, path)
+                        .or_else(|| find_fields_from_term_kind(linearization, id, path))
+                        .unwrap_or_default()
                 }
-                _ => None,
+                (
+                    TermKind::RecordField {
+                        value: ValueState::Known(value),
+                        ..
+                    },
+                    _,
+                ) => find_fields_from_term_kind(linearization, *value, path).unwrap_or_default(),
+                _ => Vec::new(),
             }
         })
-        .collect()
+        .unwrap_or_default()
 }
 
 /// Generate possible completion identifiers given a source text, its linearization
 /// and the current item the cursor points at.
 fn get_completion_identifiers(
     source: &str,
+    file: FileId,
     trigger: Option<&str>,
     linearization: &Completed,
     item: &LinearizationItem<Types>,
+    server: &Server,
 ) -> Result<Vec<CompletionItem>, ResponseError> {
+    fn complete(
+        item: &LinearizationItem<Types>,
+        name: Ident,
+        linearization: &Completed,
+        server: &Server,
+        path: &mut Vec<Ident>,
+        file: FileId,
+    ) -> Option<Vec<Ident>> {
+        item.env
+            .get(&(name, file))
+            .map(|key| (key, linearization))
+            .or_else(|| {
+                // If we can't find in the file's scope, check it from the stdlib's scope.
```

Comment:
Here, if you had `Environment = <Ident, (FileId, usize)>`, you wouldn't have to do this `or_else` special casing.